src/read.jl

```diff
@@ -261,3 +255,197 @@ function read_demand_flexibility(filepath, interval)::DemandFlexibility
+    data to bus data
+"""
+function reformat_demand_flexibility_input(
+    case::Case, demand_flexibility::DemandFlexibility, sets::Sets
+)
+    # list of zones in the network
+    zone_list = sort(collect(Set(case.bus_zone)))
+    # distribute zone-level aggregated number to buses
+    zone_to_bus_shares = _make_bus_demand_weighting(case)
+    # incidence matrix of mapping between zone and bus
+    zone_to_bus_incidence = deepcopy(zone_to_bus_shares)
+    (x, y, v) = findnz(zone_to_bus_incidence)
+    for i in 1:nnz(zone_to_bus_incidence)
+        # buses with no load still show up as nonzero entries
+        if v[i] > 0
+            zone_to_bus_incidence[x[i], y[i]] = 1
+        end
+    end
+
+    # create an empty demand flexibility object and copy unchanged fields
+    demand_flexibility_updated = Dict(
+        "duration" => demand_flexibility.duration,
+        "enabled" => demand_flexibility.enabled,
+        "interval_balance" => demand_flexibility.interval_balance,
+        "rolling_balance" => demand_flexibility.rolling_balance,
+        "flex_amt_up" => nothing,
+        "flex_amt_dn" => nothing,
+        "cost_up" => nothing,
+        "cost_dn" => nothing,
+    )
+
+    # iterate through fields
+    for field in ["flex_amt_up", "flex_amt_dn", "cost_up", "cost_dn"]
+        # dataframe in the corresponding field
+        fh = getfield(demand_flexibility, Symbol(field))
+
+        # skip if no cost
+        if isnothing(fh)
+            continue
+        end
+
+        # headers
+        flexible_str = names(fh)[2:end]
+
+        # index of columns specifing the flexibility of a zone
+        zone_columns_idx = findall(x -> occursin("zone.", x), flexible_str)
+
+        # numeric ID corresponding to zone columns
+        flexible_zone_id = [
+            parse(Int64, replace(flexible_str[i], "zone." => "")) for i in zone_columns_idx
+        ]
+        flexible_zone_num = length(flexible_zone_id)
+
+        # do zone-bus conversion only when zone columns are present
+        if flexible_zone_num > 0
+
+            # index of columns specifing the flexibility of a bus
+            bus_columns_idx = findall(x -> !occursin(".", x), flexible_str)
+
+            # numeric ID corresponding to bus columns
+            flexible_bus_id = [parse(Int64, flexible_str[i]) for i in bus_columns_idx]
+            flexible_bus_num = length(flexible_bus_id)
+
+            # check if zone numbers are correct
+            if !all([issubset(i, zone_list) for i in flexible_zone_id])
+                @error("Invalid load zone numeric ID(s) in demand flexibility input files!")
+            elseif !all([issubset(i, case.busid) for i in flexible_bus_id])
+                @error("Invalid load bus numeric ID(s) in demand flexibility input files!")
+            end
+
+            # list index of each zone in input file in the sorted list of zones
+            zone_cols_idx = findall(x -> x in flexible_zone_id, zone_list)
```

Comment:
this is the index in the zone list (which will be used later to index the zone-bus mapping matrix). e.g. if there are total 3 zones [1, 2, 3]; 3 zone columns [2, 3, 1] then this index would be [1, 2, 0]

Reply:
Hmm when I test this they seem to produce the same array, even when the zone columns are not ordered. For an example with four zones (`["zone.301", "zone.302", "zone.304", "zone.303"]`), I get that both `zone_columns_idx` and `zone_cols_idx` are equal to `[1, 2, 3, 4]`.

Reply:
Just confirmed the problem.. It seems this findall iteration is not keeping the original order of elements in flexible_zone_id. I had to use `zone_cols_idx = [findfirst(y -> y == x , zone_list) for x in flexible_zone_id]` to avoid losing the order. This behavior of search related fcns in Julia is so strange..